### PR TITLE
Add Ruff

### PR DESCRIPTION
**Pull request by davidotte**
### Change Description

**Brief** (a few bullet points describing your changes, use full sentences and try to link lines in the code whenever needed)

Add ruff and pre-commit to client repo (analogous to server repo)

**Details** (add details if your pull request is more complicated and harder to understand from the code alone)

### Standard Qs (leave questions that do not apply blank)
**If you broke behavior: Please describe what behavior you broke and how you inform people to not get stuck trying to use the old behavior.**

**If you used new dependencies: Did you add them to `requirements.txt`?**

yes

**Who did you ping on Mattermost to review your PR? Please ping that person again whenever you are ready for another review.**

@liam-sbhoo 

---
*Please do not mark comments/conversations as resolved unless you are the assigned reviewer. This helps maintain clarity during the review process.*

### CODE REVIEW

File: .github/workflows/pull_request.yml [hunk 1]
```diff
@@ -5,6 +5,16 @@ on:
       - main
 
 jobs:
+  check_python_linting:
+    name: Ruff Linting
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
```
Comment (liam-sbhoo) [inline .github/workflows/pull_request.yml:12]: Maybe use v4? That's the latest version available and `tabpfn-server` is using v4 too.

File: .github/workflows/pull_request.yml [hunk 2]
```diff
@@ -5,6 +5,16 @@ on:
...
+      - uses: chartboost/ruff-action@v1
+        with:
```
Comment (liam-sbhoo) [inline .github/workflows/pull_request.yml:14]: Could you confirm if the current GitHub workflow check for formatting issues? I have a vague memory that our github test only checks for linting issues.

This used to be a bug of ruff-action package, and it seems they finally fixed it here (https://github.com/ChartBoost/ruff-action?tab=readme-ov-file#use-ruff-format).

File: README.md [hunk 3]
```diff
@@ -43,3 +43,10 @@ tabpfn.fit(X_train, y_train)
 tabpfn.predict(X_test)
 # or you can also use tabpfn.predict_proba(X_test)
 ```
+
+# Development
```
Comment (liam-sbhoo) [inline README.md:47]: Can you also mention this (https://docs.astral.sh/ruff/integrations/) to encourage developers to have the extensions installed in the editor?

**Review by liam-sbhoo — CHANGES_REQUESTED** [review]
Thank you for picking this up! Just some minor comments.

File: .github/workflows/pull_request.yml (hunk 2 above)
Comment (davidotte) [reply .github/workflows/pull_request.yml:14]: Okay yeah, it looks like we have to add this argument such that formatting is checked. But it seems like as soon I add the argument, linting is not checked any longer, at least the output logs suggest that. So I just added it with and without the argument :thinking:

**Pull request merged**